File: swift/tests/WalletKitTests/SimpleTest.swift
import XCTest
@testable import WalletKit

private final class CapturingLogger: WalletKit.Logger {
    private let lock = NSLock()
    private var entries: [(WalletKit.LogLevel, String)] = []

    func log(level: WalletKit.LogLevel, message: String) {
        lock.lock()
        entries.append((level, message))
        lock.unlock()
    }

    func snapshot() -> [(WalletKit.LogLevel, String)] {
        lock.lock()
        defer { lock.unlock() }
        return entries
    }
}

final class SimpleTest: XCTestCase {
    func testInitLoggingForwardsLevelAndMessage() {
        let logger = CapturingLogger()
        WalletKit.initLogging(logger: logger, level: .info)
        WalletKit.emitLog(level: .info, message: "bridge test")

        // Log delivery happens on a dedicated background thread via an mpsc
        // channel. Poll with a generous timeout so CI runners under load
        // don't flake.
        let deadline = Date().addingTimeInterval(5.0)
        var entries: [(WalletKit.LogLevel, String)] = []
        while Date() < deadline {
            entries = logger.snapshot()
            if entries.contains(where: { $0.0 == .info && $0.1.contains("bridge test") }) {
                break
            }
            Thread.sleep(forTimeInterval: 0.05)
        }


        XCTAssertFalse(entries.isEmpty, "expected at least one bridged log entry")

        let hasBridgedMessage = entries.contains { level, message in
            level == .info && message.contains("bridge test")
        }
        XCTAssertTrue(hasBridgedMessage, "expected info-level bridged log")
    }
}

File: kotlin/walletkit-tests/src/test/kotlin/org/world/walletkit/SimpleTest.kt
package org.world.walletkit

import uniffi.walletkit_core.FieldElement
import uniffi.walletkit_core.LogLevel
import uniffi.walletkit_core.Logger
import uniffi.walletkit_core.OwnershipProof
import uniffi.walletkit_core.emitLog
import uniffi.walletkit_core.initLogging
import kotlin.test.Test
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

private class CapturingLogger : Logger {
    private val lock = Any()
    private val entries = mutableListOf<Pair<LogLevel, String>>()

    override fun log(
        level: LogLevel,
        message: String,
    ) {
        synchronized(lock) {
            entries.add(level to message)
        }
    }

    fun snapshot(): List<Pair<LogLevel, String>> =
        synchronized(lock) {
            entries.toList()
        }
}

class SimpleTest {
    @Test
    fun initLoggingForwardsLevelAndMessage() {
        val logger = CapturingLogger()
        initLogging(logger, LogLevel.INFO)
        emitLog(LogLevel.INFO, "bridge test")

        // Log delivery happens on a dedicated background thread, so give it
        // a moment to flush through the channel.
        Thread.sleep(50)

        val entries = logger.snapshot()
        assertTrue(entries.isNotEmpty(), "expected at least one bridged log entry")

        val hasBridgedMessage =
            entries.any { (level, message) ->
                level == LogLevel.INFO && message.contains("bridge test")
            }
        assertTrue(hasBridgedMessage, "expected info-level bridged log")
    }

    // Compile-time only: a rename, a dropped export, or a changed parameter
    // type fails here. Calling it needs a proof, so that stays in Rust.
    @Test
    fun ownershipProofExposesVerificationRequestBuilder() {
        fun callBuilder(
            proof: OwnershipProof,
            challengeId: String,
            credentialSub: FieldElement,
        ): String = proof.toVerificationRequestJson(challengeId, credentialSub)

        assertNotNull(::callBuilder)
    }
}
